patches_for_WebRTC_org/m80/0007.5-Porting-H264-encoder-from-the-WebRTC-UWP-project.patch

```diff
@@ -0,0 +1,3098 @@
++  ON_SUCCEEDED(mediaTypeIn->SetGUID(MF_MT_SUBTYPE, MFVideoFormat_NV12));
++  ON_SUCCEEDED(mediaTypeIn->SetUINT32(MF_MT_INTERLACE_MODE,
++                                      MFVideoInterlace_Progressive));
++  ON_SUCCEEDED(mediaTypeIn->SetUINT32(MF_MT_ALL_SAMPLES_INDEPENDENT, TRUE));
++  ON_SUCCEEDED(
++      MFSetAttributeSize(mediaTypeIn.Get(), MF_MT_FRAME_SIZE, width_, height_));
++  ON_SUCCEEDED(MFSetAttributeRatio(mediaTypeIn.Get(), MF_MT_FRAME_RATE,
++                                   max_frame_rate_, 1));
++
++  // Create the media sink
++  ON_SUCCEEDED(Microsoft::WRL::MakeAndInitialize<H264MediaSink>(
++      mediaSink_.ReleaseAndGetAddressOf()));
++
++  // SinkWriter creation attributes
++  ON_SUCCEEDED(MFCreateAttributes(
++      sinkWriterCreationAttributes_.ReleaseAndGetAddressOf(), 1));
++  ON_SUCCEEDED(sinkWriterCreationAttributes_->SetUINT32(
++      MF_READWRITE_ENABLE_HARDWARE_TRANSFORMS, TRUE));
++  ON_SUCCEEDED(sinkWriterCreationAttributes_->SetUINT32(
++      MF_SINK_WRITER_DISABLE_THROTTLING, TRUE));
++  ON_SUCCEEDED(sinkWriterCreationAttributes_->SetUINT32(MF_LOW_LATENCY, TRUE));
++
++  // Create the sink writer
++  ON_SUCCEEDED(MFCreateSinkWriterFromMediaSink(
++      mediaSink_.Get(), sinkWriterCreationAttributes_.Get(),
++      sinkWriter_.ReleaseAndGetAddressOf()));
++
++  // Add the h264 output stream to the writer
++  ON_SUCCEEDED(sinkWriter_->AddStream(mediaTypeOut.Get(), &streamIndex_));
++
++  // SinkWriter encoder properties
++  ON_SUCCEEDED(MFCreateAttributes(
++      sinkWriterEncoderAttributes_.ReleaseAndGetAddressOf(), 1));
++  ON_SUCCEEDED(
++      sinkWriter_->SetInputMediaType(streamIndex_, mediaTypeIn.Get(), nullptr));
++
++  // Register this as the callback for encoded samples.
++  ON_SUCCEEDED(mediaSink_->RegisterEncodingCallback(this));
++
++  ON_SUCCEEDED(sinkWriter_->BeginWriting());
++
++  codec_ = *codec_settings;
++
++  if (SUCCEEDED(hr)) {
++    inited_ = true;
++    lastTimeSettingsChanged_ = rtc::TimeMillis();
++    return WEBRTC_VIDEO_CODEC_OK;
++  } else {
++    return hr;
++  }
++}
++
++//
++int32_t H264EncoderMFImpl::InitEncode(const VideoCodec* inst,
++                                      const VideoEncoder::Settings& settings) {
++  if (!inst || inst->codecType != kVideoCodecH264) {
++    RTC_LOG(LS_ERROR) << "H264 UWP Encoder not registered as H264 codec";
++    return WEBRTC_VIDEO_CODEC_ERR_PARAMETER;
++  }
++  if (inst->maxFramerate == 0) {
++    RTC_LOG(LS_ERROR) << "H264 UWP Encoder has no framerate defined";
++    return WEBRTC_VIDEO_CODEC_ERR_PARAMETER;
++  }
++  if (inst->width < 1 || inst->height < 1) {
++    RTC_LOG(LS_ERROR) << "H264 UWP Encoder has no valid frame size defined";
++    return WEBRTC_VIDEO_CODEC_ERR_PARAMETER;
++  }
++
++  width_ = inst->width;
++  height_ = inst->height;
++  target_bps_ = inst->startBitrate > 0 ? inst->startBitrate * 1000
++                                       : width_ * height_ * 2.0;
++
```

Comment:
Where are the latest changes for setting maxQp?

https://github.com/webrtc-uwp/webrtc-windows/blob/1f59c3096c009bdc8d86169fd00e5950def19c82/third_party/winuwp_h264/H264Encoder/H264Encoder.cc#L117